command/agent/retry_join.go

```diff
@@ -0,0 +1,90 @@
+package agent
+
+import (
+	"log"
+	"strings"
+	"time"
+)
+
+// DiscoverInterface is an interface for the Discover type in the go-discover
+// library. Using an interface allows for ease of testing.
+type DiscoverInterface interface {
+	// Addrs discovers ip addresses of nodes that match the given filter
+	// criteria.
+	// The config string must have the format 'provider=xxx key=val key=val ...'
+	// where the keys and values are provider specific. The values are URL
+	// encoded.
+	Addrs(string, *log.Logger) ([]string, error)
+
+	// Help describes the format of the configuration string for address
+	// discovery and the various provider specific options.
+	Help() string
+
+	// Names returns the names of the configured providers.
+	Names() []string
+}
+
+// retryJoiner is used to handle retrying a join until it succeeds or all of
+// its tries are exhausted.
+type retryJoiner struct {
+	// join adds the specified servers to the serf cluster
+	join func([]string) (int, error)
+
+	// discover is of type Discover, where this is either the go-discover
+	// implementation or a mock used for testing
+	discover DiscoverInterface
+
+	// errCh is used to communicate with the agent when the max retry attempt
+	// limit has been reached
+	errCh chan struct{}
+
+	// logger is the agent logger.
+	logger *log.Logger
+}
+
+// retryJoin is used to handle retrying a join until it succeeds or all retries
+// are exhausted.
+func (r *retryJoiner) RetryJoin(config *Config) {
+	if len(config.Server.RetryJoin) == 0 || !config.Server.Enabled {
+		return
+	}
+
+	attempt := 0
+
+	r.logger.Printf("[INFO] agent: Joining cluster...")
+
+	for {
+		var addrs []string
+
+		for _, addr := range config.Server.RetryJoin {
+			switch {
+			case strings.Contains(addr, "provider"):
```

Comment:
Could you add a comment describing why we need to do this strings.Contains call?

Will this break for anyone with "provider" anywhere in a server name? That seems unlikely, and perhaps there's no better heuristic?

Reply:
Good point- if it is a domain name with "provider", this might break. I can add more logic to guard against this. 

See here for all accepted inputs on Consul (which we are looking for feature parity) 
https://www.consul.io/docs/agent/options.html#retry-join

Reply:
I suppose we could have a check instead for "provider=" to test that this is the start to a go-provider cloud metadata entry, and (i believe) "=" won't be part of any domain name.

Reply:
Using `provider=` seems perfect!